neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
+
+        for current_id, data_input, expected_output in data_container:
+            checkpoint_current_id, checkpoint_data_input, checkpoint_expected_outputs = \
+                self.data_checkpointer.read_current_id_checkpoint(
+                    context.get_path(),
+                    current_id
+                )
+
+            list_data_container.append(
+                checkpoint_current_id,
+                checkpoint_data_input,
+                checkpoint_expected_outputs
+            )
+
+        return list_data_container
 
     def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
         """
-        Whether or not we should resume the pipeline (if the checkpoint exists)
+        Returns True if the checkpoint exists for each data inputs.
 
         :param context: execution context
-        :param data_container: data to resume
-        :return:
-        """
-        self.set_checkpoint_path(context.get_path())
-        return self._checkpoint_exists(data_container)
-
-    def _checkpoint_exists(self, data_container: DataContainer) -> bool:
-        """
-        Returns True if the checkpoints for each data input id exists
         :param data_container:
         :return:
         """
         for current_id in data_container.current_ids:
-            if not os.path.exists(self.get_checkpoint_file_path(current_id)):
+            if not os.path.exists(
+                    self.data_checkpointer.get_checkpoint_file_path(
+                        context.get_path(),
+                        current_id
+                    )
+            ):
                 return False
 
         return True
 
-    def get_checkpoint_file_path(self, current_id) -> str:
+
+class Checkpoint(ResumableMixin, BaseStep):
+    """
+    Resumable Checkpoint Step to load, and save both data checkpoints, and step checkpoints.
+    Checkpoint uses a list of step checkpointers(List[StepCheckpointer]), and data checkpointers(List[BaseCheckpointer]).
+
+    Data Checkpoints save the state of the data container (transformed data inputs, and expected outputs)
+    for the current execution mode (fit, fit_transform, or transform).
+
+    Step Checkpoints save the state of the fitted steps before the checkpoint
+    for the current execution mode (fit or fit_transform).
+
+    By default(no arguments specified), the Checkpoint step saves the step checkpoints for any fit or fit transform,
+    and saves a different data checkpoint with pickle data container checkpointers :
+
+    ```
+    Checkpoint(
+        step_checkpointers=[
+            StepCheckpointer(ExecutionMode.FIT_OR_FIT_TRANSFORM)
+        ]
+        data_checkpointers=[
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.TRANSFORM),
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT),
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT_TRANSFORM),
+        ]
```

Comment:
before reading this code, I thought we could put all checkpointers in the same list. What you did might be okay too compared to my expectations, this is a fast code revirew.

Reply:
it didn't feel right to have step checkpointers, and data checkpointers in the same list. it is two different things

Reply:
Don't they inherit from the same class (and hence have all the same method signatures)? Is there something that forces us to treat them separately?

Reply:
(I guess yes, in that case you're okay)